src/dataflow/flow-graph.ts

```diff
@@ -53,6 +55,95 @@ export class FlowGraph {
     }
     return connections;
   }
+
+  /** Returns true if all checks in the graph pass. */
+  validateGraph(): boolean {
+    for (const edge of this.edges) {
+      if (edge.check) {
+        const success = this.validateSingleEdge(edge);
+        if (!success) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  /** Validates a single check (on the given edge). Returns true if the check passes. */
+  private validateSingleEdge(edge: Edge): boolean {
+    if (!edge.check) {
+      throw new Error('Edge does not have any check conditions.');
+    }
+
+    const check = edge.check;
+    const startPath = BackwardsPath.fromEdge(edge);
+    const stack: BackwardsPath[] = [startPath];
+
+    while (stack.length) {
+      const path = stack.pop();
+      const nodeToCheck = path.endNode;
+      // TODO: Check the node itself too. The node might be untagged (e.g. if it has no input edges), so it could fail a check.
+      const edgesToCheck = nodeToCheck.inEdges;
+      for (const edge of edgesToCheck) {
+        const result = edge.isCheckSatisfied(check);
+        switch (result) {
+          case CheckResult.Success:
+            // Check was met. Continue checking other paths.
+            continue;
+          case CheckResult.Failure:
+            // Check failed.
+            return false;
+          case CheckResult.KeepGoing:
+            // Check has not failed yet for this path yet. Add it to the stack and keep going.
+            stack.push(path.appendEdge(edge));
+            continue;
+          default:
+            throw new Error(`Unknown check result: ${result}`);
+        }
+      }
+    }
+    return true;
+  }
+}
+
+export enum CheckResult {
+  Success,
+  Failure,
+  KeepGoing,
+}
+
+/**
+ * A path that walks backwards through the graph, i.e. it walks along the directed edges in the reverse direction. The path is described by the
+ * nodes in the path. 
+ */
+export class BackwardsPath {
+  private constructor(readonly nodes: Node[]) {}
+
+  static fromEdge(edge: Edge) {
+    return new BackwardsPath([edge.end, edge.start]);
+  }
+
+  appendEdge(edge: Edge): BackwardsPath {
```

Comment:
I see that BackwardsPath is immutable, and you make use of this in the graph check; this is a good design choice overall, but a couple of improvements I'd like to see:
 - comment that BackwardsPath is immutable to make it clear to readers of the code.
 - find a name for appendEdge that doesn't imply it acts on a mutable structure.

Reply:
done!